### PR TITLE
Laboratory work #2, Sviridov Vyacheslav - 20FPL1

### PR DESCRIPTION
**Pull request by VSvir**

### CODE REVIEW

File: lab_2/main.py [hunk 1]
```diff
@@ -13,7 +13,20 @@ def get_freq_dict(tokens: list) -> dict or None:
...
+    for token in tokens:
+        if not isinstance(token, str):
+            return None
+        if token not in freq_dict:
+            freq_dict[token] = 1
+        else:
+            freq_dict[token] += 1
+    for key, value in freq_dict.items():
+        value = round(value / len(tokens), 5)
+        freq_dict[key] = value
```
Comment (ValeriyaKuznetsova) [inline lab_2/main.py:28]: can you join this and the previous line?

File: lab_2/main.py [hunk 2]
```diff
@@ -24,7 +37,15 @@ def get_language_profiles(texts_corpus: list, language_labels: list) -> dict or
...
     """
-    pass
+    if not isinstance(texts_corpus, list) or not isinstance(language_labels, list):
+        return None
+    language_profiles = {}
+    for text in texts_corpus:
+        if not isinstance(text, list):
+            return None
+        freq_dict = get_freq_dict(text)
+        language_profiles[language_labels[texts_corpus.index(text)]] = freq_dict
```
Comment (ValeriyaKuznetsova) [inline lab_2/main.py:47]: what happens here?

File: lab_2/main.py [hunk 3]
```diff
@@ -43,7 +72,23 @@ def get_text_vector(original_text: list, language_profiles: dict) -> list or Non
...
+    words_freq_dict = {}
+    for dicts in words_freq:
+        for key, value in dicts.items():
+            if key in words_freq_dict:
+                if words_freq_dict.get(key) > dicts[key]:
+                    continue
+            words_freq_dict[key] = value
+    for unique_word in get_language_features(language_profiles):
+        if unique_word in original_text:
+            text_vector.append(words_freq_dict.get(unique_word))
```
Comment (ValeriyaKuznetsova) [inline lab_2/main.py:88]: there is a way to get a value if a key exists or 0 if it doesn't

File: lab_2/main.py [hunk 4]
```diff
@@ -53,7 +98,17 @@ def calculate_distance(unknown_text_vector: list, known_text_vector: list) -> fl
...
+    if not isinstance(unknown_text_vector, list) or not isinstance(known_text_vector, list):
+        return None
+    distance = 0
+    counter = 0
+    for vector_value in unknown_text_vector:
+        if not isinstance(vector_value, (int, float)) or not \
+                isinstance(known_text_vector[counter], (int, float)):
+            return None
+        distance += (vector_value - known_text_vector[counter]) ** 2
+        counter += 1
```
Comment (ValeriyaKuznetsova) [inline lab_2/main.py:110]: maybe you can use `index` on the `for` line instead of a counter?

File: lab_2/main.py [hunk 5]
```diff
@@ -89,8 +168,42 @@ def predict_language_knn(unknown_text_vector: list, known_text_vectors: list,
...
     """
-    pass
-
+    if not isinstance(unknown_text_vector, list) or not \
+        isinstance(known_text_vectors, list) or not \
+            isinstance(language_labels, list) or not \
+            len(known_text_vectors) == len(language_labels):
+        return None
+    if metric == 'euclid':
+        vectors_distances = []
```
Comment (ValeriyaKuznetsova) [inline lab_2/main.py:177]: seems like you need these lines in any case, I can see duplicate code

File: lab_2/main.py [hunk 6]
```diff
@@ -89,8 +168,42 @@ def predict_language_knn(unknown_text_vector: list, known_text_vectors: list,
...
+            if not isinstance(known_text_vector, list):
+                return None
+            vectors_distances.append(round(calculate_distance_manhattan(unknown_text_vector,
+                                                                        known_text_vector), 5))
+    else:
+        return None
+    sorted_vectors_distances = vectors_distances.copy()
+    sorted_vectors_distances.sort()
+    min_distance_vectors = sorted_vectors_distances[:k]
+    for min_distance_vector in min_distance_vectors:
```
Comment (ValeriyaKuznetsova) [inline lab_2/main.py:198]: better to get index here

**Review by ValeriyaKuznetsova — COMMENTED** [review]
Good one, some comments to consider

Don't forget about `start.py`

File: lab_2/main.py (hunk 6 above)
Comment (VSvir) [reply lab_2/main.py:198]: Can you explain what you mean, please? I can't think of a way I could get index in `for` line or especially before of it

File: lab_2/main.py (hunk 4 above)
Comment (VSvir) [reply lab_2/main.py]: Had to replace `range` with `enumerate`

**Pull request closed without merging**